### ISSUE TITLE
2D dimension spike train dissmillarity cannot be calculated directly

### ISSUE THREAD

**Issue by tasbolat1** [NeuralEnsemble/elephant#309]
Hello elephant community,

I have been trying to compute spike train dissimilarity using VP and VR distances for 2D data (in my case $156 \times 325$, 325 is time dimension, 156 is spatial dimension), but it always gives me error saying that it cannot broadcast. is it possible to compute the spike train dissimilarities for 2d or higher dimensions?

Thanks in advance

**Comment by mdenker** [comment]
Hi, I'm not sure I fully understand what is meant by VP and VR on 2-D data. 

The current implementation take in lists of Neo `Spiketrain` objects, which contain the spike times of individual neurons, independent of spatial locations (https://elephant.readthedocs.io/en/latest/reference/spike_train_dissimilarity.html).

When you say 2D data, are you referring to a 0/1 matrix of (time x neuronID)? In that case the matrix would have to be converted to Spiketrain objects first -- in fact that's an interesting use case that we didn't come across yet, and probably there should be a function for this -- I'll put that as an issue. For now, you'd have to do something like (no guarantee):

```python
binning = 1 * pq.ms  # time bin size of the matrix
spiketrain_list = []
for neuron_id in range(matrix.shape[1]):
    spiketimes = np.where(matrix[:,neuron_id]==1)(0) * binning
    spiketrain_list.append(
        neo.Spiketrain(spiketimes, t_start=0*pq.ms, t_stop=matrix.shape[0] * binning))

...then use spiketrain_list....
```

**Comment by tasbolat1** [comment]
@mdenker thanks for promt reply,
I meant VP -  Victor–Purpura and VR - van Rossum distances. Let me clarify my problem in more detail. I have two spike trains X1 and X2. Both consist of 156 neurons and it's defined by `neo.Spiketrain(.)`. The problem is that I can only calculate distances between neurons, but I want to measure distance between `X1` and `X2`. So, basically, the times argument of `neo.Spiketrain` consists 2D matrix, rather list of values.

**Comment by mdenker** [comment]
So, to understand correctly, `X1` is a Neo `Spiketrain` object, but `X.times` is a 2D matrix, where dimension one goes through spike times and dimension 2 through 156 neurons? 

As an example, if neuron 1 spikes at 1 ms and 5 ms and neuron 2 spikes at 6 ms and 10 ms, the `.times` matrix would look like the following?
```python
[ 1, 6;
5, 10] * pq.ms
```

If this is the case, the problem is that there should be exactly one `Spiketrain` per neuron (and per trial/condition/time stretch/...), i.e., `.times` must but a vector. Having multiple neurons per `Spiketrain` object is problematic, since the spike counts are in general not equal (as in the mock example above, where both neurons fire exactly 2 spikes). Therefore you cannot in general specify a 2D matrix, unless you would introduce something like padding with `None` or so (a workaround for this is currently in dev in Neo for a container object to hold multiple spiketrains).

So the solution would be to have a list of Spiketrain objects, all of them carrying one spiketrain (with a list of its spike times). Hope this helps?

**Comment by JuliaSprenger** [comment]
I agree with @mdenker, that the times of a spiketrain should be a 1d vector for exactly the reasons you listed. Unfortunately, this was not strictly enforced by `neo` yet, but will (hopefully) be changed soon (https://github.com/NeuralEnsemble/python-neo/pull/802). @tasbolat1 Does this clarify the issue?

**Comment by dizcza** [comment]
@tasbolat1 is your 2D matrix is a binary matrix of zeros and ones?

**Comment by tasbolat1** [comment]
Thanks everyone. @dizcza, @mdenker, the `times` is 2D matrix.  @JuliaSprenger, yes it clarifies everything.  Finally, I did it using add-hoc method, basically looping over each row. However, it would be useful if one can calculate spike train dissimilarities for multi-neurons.

**Comment by dizcza** [comment]
Yeah, currently we support what neo supports. As soon as neo releases SpikeTrainList object that stores 2D matrix of times padded with NaNs, we start adding its support in Elephant as well. For now, the only solution is the add-hoc method that you mentioned.

If you want to speed things up, I suggest using our [parallel](https://github.com/NeuralEnsemble/elephant/tree/master/elephant/parallel) module ([docs](https://elephant.readthedocs.io/en/latest/reference/parallel.html), [tutorial](https://elephant.readthedocs.io/en/latest/tutorials/parallel.html)). It's not yet in the pip release, so you need to install the development version from the master branch (see [Installation](https://elephant.readthedocs.io/en/latest/install.html#id2), choose Development version).

I'll leave it open for a while to see any discussion ongoing and close it in a week because, of course, we don't support incorrect ways of using neo data structures.

**Comment by mdenker** [comment]
Ok, this clarifies it. However, then the issue is more about binary spike trains vs spike-time representations. I will  think about #310. This would enable to do something roughly like
```
a= [...matrix of ones and zeros)
spiketrain_list = elephant.conv.BinnedSpiketrain(a).to_spiketrain_list()
result = elephant.spiketrain_dissimilarity.victor_pupura(spiketrain_list)
```

**Comment by dizcza** [comment]
`BinnedSpikeTrain.to_spike_trains()` feature request (make sense when your input data is a binned 2D matrix) is implemented in #390
Feel welcome to try out Michael's suggestion and share the feedback/bugs in a new issue.
Thanks.
